File: epi_judge_python/pivot_list.py
import functools
from typing import Optional

from list_node import ListNode
from test_framework import generic_test
from test_framework.test_failure import TestFailure
from test_framework.test_utils import enable_executor_hook

def prnt(l):
    pt = l
    print()
    while pt.next:
        print(pt.data, ' ', end="")
        pt = pt.next
    print(pt.data, ' ', end="")
    print()


def list_pivoting(l: ListNode, x: int) -> Optional[ListNode]:
    if not l or not l.next:
        return l
    dummyh = ListNode(0, l)
    piv = prev = l
    while piv and piv.data != x and piv.next:
        prev = piv
        piv = piv.next
    last = piv
    while last.next:
        last = last.next
    it, it2 = l, dummyh
    while it and it.next and it is not piv:
        if it.data > x:
            it2.next = it.next
            last.next = it
            last = last.next
            last.next = None
            it = it2.next

        else:
            it2 = it
            it = it.next

    it = it2 = piv
    while it is not last and it.next:
        if it.data < x:
            it2.next = it.next
            it.next = piv
            prev.next = it
            prev = prev.next
            it = it2.next
        else:
            it2 = it
            it = it.next
    prnt(dummyh.next)
    prnt(l)
    return l 


def linked_to_list(l):
    v = list()
    while l is not None:
        v.append(l.data)
        l = l.next
    return v


@enable_executor_hook
def list_pivoting_wrapper(executor, l, x):
    original = linked_to_list(l)

    l = executor.run(functools.partial(list_pivoting, l, x))

    pivoted = linked_to_list(l)
    mode = -1
    for i in pivoted:
        if mode == -1:
            if i == x:
                mode = 0
            elif i > x:
                mode = 1
        elif mode == 0:
            if i < x:
                raise TestFailure('List is not pivoted')
            elif i > x:
                mode = 1
        else:
            if i <= x:
                raise TestFailure('List is not pivoted')

    if sorted(original) != sorted(pivoted):
        raise TestFailure('Result list contains different values')


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('pivot_list.py', 'pivot_list.tsv',
                                       list_pivoting_wrapper))

File: epi_judge_cpp/pivot_list.cc
#include <algorithm>
#include <iterator>
#include <memory>
#include <vector>

#include "list_node.h"
#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"
#include "test_framework/timed_executor.h"
using std::shared_ptr;
shared_ptr<ListNode<int>> ListPivoting(const shared_ptr<ListNode<int>>& l,
                                       int x) {
  auto lesser = std::make_shared<ListNode<int>>(), lp{lesser};
  auto equal = std::make_shared<ListNode<int>>(), ep{equal};
  auto greater = std::make_shared<ListNode<int>>(), gp{greater};

  auto append = [](shared_ptr<ListNode<int>> &ptr, shared_ptr<ListNode<int>> &elem) {
    ptr->next = elem;
    ptr = ptr->next;
  };

  for (auto p = l; p; p = p->next) {
    if (p->data < x) {
      append(lp, p);
    } else if (p->data > x) {
      append(gp, p);
    } else {
      append(ep, p);
    }
  }
  gp->next.reset();
  ep->next = greater->next;
  lp->next = equal->next;
  return lesser->next;
}

std::vector<int> ListToVector(const shared_ptr<ListNode<int>>& l) {
  std::vector<int> v;
  ListNode<int>* it = l.get();
  while (it) {
    v.push_back(it->data);
    it = it->next.get();
  }
  return v;
}

void ListPivotingWrapper(TimedExecutor& executor,
                         const shared_ptr<ListNode<int>>& l, int x) {
  std::vector<int> original = ListToVector(l);

  std::shared_ptr<ListNode<int>> pivoted_list =
      executor.Run([&] { return ListPivoting(l, x); });

  std::vector<int> pivoted = ListToVector(pivoted_list);
  enum class Mode { kLess, kEq, kGreater } mode = Mode::kLess;
  for (auto& i : pivoted) {
    switch (mode) {
      case Mode::kLess:
        if (i == x) {
          mode = Mode::kEq;
        } else if (i > x) {
          mode = Mode::kGreater;
        }
        break;
      case Mode::kEq:
        if (i < x) {
          throw TestFailure("List is not pivoted");
        } else if (i > x) {
          mode = Mode::kGreater;
        }
        break;
      case Mode::kGreater:
        if (i <= x) {
          throw TestFailure("List is not pivoted");
        }
    }
  }
  std::sort(begin(original), end(original));
  std::sort(begin(pivoted), end(pivoted));
  if (original != pivoted) {
    throw TestFailure("Result list contains different values");
  }
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"executor", "l", "x"};
  return GenericTestMain(args, "pivot_list.cc", "pivot_list.tsv",
                         &ListPivotingWrapper, DefaultComparator{},
                         param_names);
}
